flink/src/main/java/org/apache/iceberg/flink/RowDataWrapper.java

```diff
@@ -0,0 +1,138 @@
+import org.apache.iceberg.types.Type;
+import org.apache.iceberg.types.Types;
+import org.apache.iceberg.util.DateTimeUtil;
+
+class RowDataWrapper implements StructLike {
+
+  private final LogicalType[] types;
+  private final PositionalGetter<?>[] getters;
+  private RowData rowData = null;
+
+  RowDataWrapper(RowType rowType, Types.StructType struct) {
+    int size = rowType.getFieldCount();
+
+    types = (LogicalType[]) Array.newInstance(LogicalType.class, size);
+    getters = (PositionalGetter[]) Array.newInstance(PositionalGetter.class, size);
+
+    for (int i = 0; i < size; i++) {
+      types[i] = rowType.getTypeAt(i);
+      getters[i] = buildGetter(types[i], struct.fields().get(i).type());
+    }
+  }
+
+  RowDataWrapper wrap(RowData data) {
+    this.rowData = data;
+    return this;
+  }
+
+  @Override
+  public int size() {
+    return types.length;
+  }
+
+  @Override
+  public <T> T get(int pos, Class<T> javaClass) {
+    if (rowData.isNullAt(pos)) {
+      return null;
+    } else if (getters[pos] != null) {
+      return javaClass.cast(getters[pos].get(rowData, pos));
+    }
+
+    Object value = RowData.createFieldGetter(types[pos], pos).getFieldOrNull(rowData);
+    return javaClass.cast(value);
+  }
+
+  @Override
+  public <T> void set(int pos, T value) {
+    throw new UnsupportedOperationException("Could not set a field in the RowDataWrapper because rowData is read-only");
+  }
+
+  private interface PositionalGetter<T> {
+    T get(RowData data, int pos);
+  }
+
+  private static PositionalGetter<?> buildGetter(LogicalType logicalType, Type type) {
+    switch (type.typeId()) {
+      case STRING:
+        return (row, pos) -> row.getString(pos).toString();
+
+      case FIXED:
+      case BINARY:
+        return (row, pos) -> ByteBuffer.wrap(row.getBinary(pos));
+
+      case UUID:
+        return (row, pos) -> UUID.nameUUIDFromBytes(row.getBinary(pos));
+
+      case DECIMAL:
+        DecimalType decimalType = (DecimalType) logicalType;
+        return (row, pos) -> row.getDecimal(pos, decimalType.getPrecision(), decimalType.getScale()).toBigDecimal();
+
+      case TIME:
+        return (row, pos) -> (long) row.getInt(pos);
+
+      case TIMESTAMP:
+        switch (logicalType.getTypeRoot()) {
+          case TIMESTAMP_WITHOUT_TIME_ZONE:
+            TimestampType timestampType = (TimestampType) logicalType;
+            return (row, pos) -> {
+              LocalDateTime localDateTime = row.getTimestamp(pos, timestampType.getPrecision()).toLocalDateTime();
+              return DateTimeUtil.microsFromTimestamp(localDateTime);
+            };
+
+          case TIMESTAMP_WITH_LOCAL_TIME_ZONE:
+            LocalZonedTimestampType lzTs = (LocalZonedTimestampType) logicalType;
+            return (row, pos) -> {
+              TimestampData timestampData = row.getTimestamp(pos, lzTs.getPrecision());
+              return timestampData.getMillisecond() * 1000 + Math.floorDiv(timestampData.getNanoOfMillisecond(), 1000);
```

Comment:
Nano of millisecond is always positive, right? In that case there is no need for floorDiv.